### PR TITLE
Improve treatment of VLAs in sizeof & params.

### PR DESCRIPTION
**Pull request by chathhorn**
@dwightguth please review.

### CODE REVIEW

File: examples/simple/violates-constraint/CV-EFNC3-bad.output [hunk 1]
```diff
@@ -1,9 +0,0 @@
-Error: CV-EFNC3
```
Comment (dwightguth) [inline examples/simple/violates-constraint/CV-EFNC3-bad.output]: can you please confirm we still catch this error under a different error code?

File: semantics/language/translation/error.k [hunk 2]
```diff
@@ -19,57 +20,37 @@ module C-TRANSLATION-ERROR
                +String "\nLine: " +String Int2String(Line)
                +String "\n")
           ~> HALT ...</k>
+          <curr-scope> S:Scope </curr-scope>
           <curr-program-loc>
                CabsLoc(File:String, Line:Int, _, _)
           </curr-program-loc>
           <result-value> _ => 1 </result-value>
-          requires notBool isCSV
+          requires notBool isCSV andBool S =/=K noEvalScope
```
Comment (dwightguth) [inline semantics/language/translation/error.k]: what exactly happens here if S ==K noEvalScope? Does it get stuck or what?

File: semantics/language/translation/typing/expr.k [hunk 3]
```diff
@@ -189,20 +187,130 @@ module C-TYPING-EXPR
...
+     rule (.K => elaborate(scope(noEvalScope, elabResult(E1)), scope(noEvalScope, elabResult(E2))))
+          ~> typeof(_ ? E1:KItem : E2:KItem)
+
+     syntax KItem ::= elabResult(K)
+     context elabResult(HOLE:KItem => reval(HOLE)) [result(RValue)]
+
+     rule <k> elabResult(V:RValue) => .K ...</k>
+          <elab> .K => V ...</elab>
+
+     rule <k> EXIT(_) => recover ...</k>
```
Comment (dwightguth) [inline semantics/language/translation/typing/expr.k]: oh I see here, never mind.

File: semantics/language/translation/env.k [hunk 4]
```diff
@@ -105,4 +106,32 @@ module C-ENV
                ListItem(L:LocalCell) => .List
           ...</block-stack>
 
+     rule recover ~> (K:KItem => .K)
+          requires K =/=K popLocals
+     rule (recover => .K) ~> popLocals
```
Comment (dwightguth) [inline semantics/language/translation/env.k]: one thing I'm not entirely clear on here though is, if the conditional expression contains an error, how does "recover" eventually lead it back to having a type?

**Comment by dwightguth** [comment]
I don't see anything obviously wrong but there's still a few minor things I'm unclear on with respect to how this diff works. I see you also did some refactoring of function definitions and the unspecified variable length array type. Is there anything else in this diff I should focus on as a change you made?

File: examples/simple/violates-constraint/CV-EFNC3-bad.output (hunk 1 above)
Comment (chathhorn) [inline examples/simple/violates-constraint/CV-EFNC3-bad.output]: This one is UB-CB1 now.

File: semantics/language/translation/env.k (hunk 4 above)
Comment (chathhorn) [inline semantics/language/translation/env.k]: It just comes back as ".K" if there's an error while calculating the type.

Comment (chathhorn) [inline semantics/language/translation/env.k]: Rather, if there's an error while elaborating.

**Comment by chathhorn** [comment]
I can't think of anything in particular. I think this makes our treatment of VLAs in params actually totally correct AFAICT.

**Pull request merged**